text/0000-build-std/1-background.md

```diff
@@ -0,0 +1,335 @@
+    `libc` provides these routines.
+  - Users often rely on weak linkage to override these symbols when required,
+    but in scenarios where weak linkage is not supported users must directly
+    turn the feature off.
+- To use sanitizers, the sanitizer runtimes from LLVM's compiler-rt need to
+  be linked against. Building of these is enabled in `bootstrap.toml`
+  ([`build.sanitizers`][bootstrap-sanitizers]) and they are
+  included in the rustup components shipped by the project.
+
+Dependencies of the standard library may use unstable or internal compiler and
+language features only when they are a dependency of the standard library.
+
+### Features
+[background-features]: #features
+
+There are a handful of features defined in the standard library crates'
+`Cargo.toml`s. There is currently no stable existing mechanism for users to
+enable or disable these features. The default set of features is determined by
+[logic in bootstrap][bootstrap-features-logic] and [the `rust.std-features`
+key in `bootstrap.toml`][bootstrap-features-toml]. The enabled features are
+often different depending on the target.
+
+It is also common for user crates to depend on the standard library (via
+`#![no_std]`) conditional on Cargo features being enabled or disabled (e.g. a
+`std` feature or if `--test` is used).
+
+### Target support
+[background-target-support]: #target-support
+
+The `std` crate's [`build.rs`][std-build.rs] checks for supported values of the
+`CARGO_CFG_TARGET_*` environment variables. These variables are akin to the
+conditional compilation [configuration options][conditional-compilation-config-options],
+and often correspond to parts of the target triple (for example,
+`CARGO_CFG_TARGET_OS` corresponds to the "os" part of a target triple - "linux"
+in "aarch64-unknown-linux-gnu"). This filtering is strict enough to distinguish
+between built-in targets but loose enough to match similar custom targets.
+
+When encountering an unknown or unsupported operating system then the
+`restricted_std` cfg is set. `restricted_std` marks the entire standard library
+as unstable, requiring `feature(restricted_std)` to be enabled on any crate that
+depends on it. There is no mechanism for users to enable the `restricted_std`
+feature on behalf of dependencies. There is also no such mechanism for `alloc`
+or `core`, only `std`.
+
+Cargo and rustc support custom targets, defined in JSON files according to an
+unstable schema defined in the compiler. On nightly, users can dump the
+target-spec-json for an existing target using `--print target-spec-json`. This
+JSON can be saved in a file, tweaked and used as the argument to `--target`. It
+is unintentional but custom target specifications can be used with `--target`
+even on stable toolchains ([rust#71009] proposes destabilising this behaviour).
+However, as custom targets do not have a pre-built standard library and so must
+use `-Zbuild-std`, their use is relegated to nightly toolchains in practice.
```

Comment:
Is `-Zbuild-std` going away as part of this PR? Or will there be one mechanism to build std for built-in targets (which works on stable) and a different mechanism to build std for custom targets (which only works on nightly)? That seems like a duplication of effort; as opposed to this new mechanism having an extra (unstable) flag to allow the building of custom targets. Or maybe I misunderstood.

Reply:
-Zbuild-std will be going away, but -Zbuild-std-features will stay until stage 1b is available on nightly